peer/peerdiscovery.go

```diff
@@ -0,0 +1,173 @@
+package peer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type SignatoryAndAddress struct {
+	Signatory id.Signatory
+	Address   wire.Address
+}
+
+func (sigAndAddr SignatoryAndAddress) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := sigAndAddr.Signatory.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal address: %v", err)
+	}
+	return buf, rem, err
+}
+func (sigAndAddr *SignatoryAndAddress) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := (&sigAndAddr.Signatory).Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal address: %v", err)
+	}
+	return buf, rem, err
+}
+
+type DiscoveryClient struct {
+	opts DiscoveryOptions
+
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) DiscoverPeers(ctx context.Context) {
+	var maxExpectedPeers int
+	if dc.opts.MaxExpectedPeers <= 0  {
+		dc.opts.Logger.Debug("pinging", zap.String("max expected peers cannot be less than 1", "setting to default - 5"))
+		maxExpectedPeers = 5
+	} else {
+		maxExpectedPeers = dc.opts.MaxExpectedPeers
+	}
+
+	go func() {
+		var pingData [16]byte
+		binary.LittleEndian.PutUint64(pingData[:8], uint64(maxExpectedPeers))
+		binary.LittleEndian.PutUint64(pingData[8:], uint64(dc.transport.Port()))
+
+		msg := wire.Msg{
+			Version: wire.MsgVersion1,
+			Type:    wire.MsgTypePing,
+			Data:    pingData[:],
+		}
+		for {
+			for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+				msg.To = id.Hash(sig)
+				err := dc.transport.Send(ctx, sig, msg)
+				if err != nil {
+					dc.opts.Logger.Debug("pinging", zap.Error(err))
+					dc.transport.Table().DeletePeer(sig)
+				}
+			}
+			<-time.After(dc.opts.PingTimePeriod)
```

Comment:
You should select on the context here, because it is reasonable to expect that PingTimePeriod is very long.

```
		select {
		case <-ctx.Done():
			return
		case <-time.After(dc.opts.PingTimePeriod):
		}
```

Reply:
Also, consider the different between absolute and relative delays. You have a relative delay here (no matter how long the actual pinging takes, you will wait an additional PingTimePeriod). If you want an absolute time delay (always ping exactly once per PingTimePeriod), then you should use a ticker.